Adds Azure DevOps guidance and updates GitHub Actions guidance. Closes #5879 (#6663)
pnp/cli-microsoft365: docs/docs/user-guide/azuredevops-pipeline.mdx

The first thing that we will need is a yaml file that will define our pipeline. This should be kept as one of the files in our repositories to which we plan to add the Azure DevOps pipeline. The file may be stored in any folder in the repository, but for Azure DevOps it is recommended to store your pipelines in `.azuredevops/pipelines` folder. Let's create a new file in this location. Let's name it `deploy-spfx-solution.yml`. So after you've done the previous steps your SPFx project folder structure should look like this:

![Project structure](../images/azure-devops-pipeline-tutorial/project-structure.png)

Now let's add the following content to the file:

```yaml
name: Deploy Solution
trigger:
  branches:
    include:
      - main
pool:
  vmImage: ubuntu-latest
variables:
  - name: CertificateBase64Encoded
    value: ""
  - name: CertificatePassword
    value: ""
  - name: EntraAppId
    value: ""
  - name: TenantId
    value: ""
  - name: SharePointBaseUrl
    value: ""
  - name: PackageName
    value: ""
stages:
  - stage: Build_and_Deploy
    jobs:
      - job: Build_and_Deploy
        steps:
          - task: NodeTool@0
            displayName: Use Node.js
            inputs:

>               versionSpec: 18.x

Author: ```suggestion
              versionSpec: 22.x
```

```

This yaml file will define a pipeline that will build, bundle, and package your SPFx solution and then deploy it to your tenant using CLI for Microsoft 365. Let's go over the file and explain what each part does:

```yaml
name: Deploy Solution
trigger:
  branches:
    include:
      - main
pool:
  vmImage: ubuntu-latest
```

The above part of the yaml file defines the name of the pipeline and the trigger that will start the pipeline. In this case, the pipeline will be triggered on every push to the `main` branch which is a quite typical use case. The last line defines the image that will be used to run the pipeline. In this case, we are using `ubuntu-latest` image. 

Now let's move a bit forward. The `variables` section of the yaml file defines the variables that will be used in the pipeline. In general, this is not the best approach to keep the variables values as part of your yaml file, but for the sake of simplicity, we will keep them here. Later on we will show you how to store the variables in Azure DevOps Library.

Next, let's go over the stages and steps of the pipeline:

```yaml
- task: NodeTool@0
  displayName: Use Node.js
  inputs:

>     versionSpec: 18.x

Author: ```suggestion
    versionSpec: 22.x
```

The above part installs the latest version of the CLI for Microsoft 365 and then executes a few commands in a script task.
The first command is of course the authentication. In this case, we will be signing in as an app using a certificate with a password. 
The next command sets the URL of the root SharePoint site for use in SPO commands. 
Then we are using the `spo app add` command to add the SPFx package to, in this case, tenant app catalog. 
The last command, `spo app deploy` deploys the app to the tenant app catalog to make it available for installation and present in the SharePoint online store.

## Adding an Azure DevOps pipeline

Adding your YAML file and committing and pushing it to your repo is not enough to make it an Azure DevOps pipeline. In order to make it a pipeline you will need to go over a few manual steps in the Azure DevOps portal.
Creating a new pipeline in Azure DevOps is a one-time operation and then every time you update the YAML definition Azure DevOps will automatically pick up the changes. In order to create a new pipeline go pipelines page in your Azure DevOps project

![Azure DevOps Navigation](../images/azure-devops-pipeline-tutorial/ado-menu.png)

> Next in the top right corner of the page click on the `New pipline` button. After that, follow the for by selecting the location of your repo, next the branch on which the YAML file is present, and then select the option to create a new pipeline based on `Existing Azure Pipelines YAML file`

Author: ```suggestion
Next in the top right corner of the page click the `New pipeline` button. After that, follow the form by selecting the location of your repo, next the branch on which the YAML file is present, and then select the option to create a new pipeline based on `Existing Azure Pipelines YAML file`
```
